pkg/compose/create.go

```diff
@@ -138,71 +158,152 @@ func (s *composeService) ensureNetworks(ctx context.Context, networks types.Netw
 	return nil
 }
 
-func (s *composeService) ensureProjectVolumes(ctx context.Context, project *types.Project) error {
+func (s *composeService) ensureProjectVolumes(ctx context.Context, project *types.Project, recreateVolumes bool) error {
+	// create observed state for volumes
+	observedVolumesState, err := s.getVolumes(ctx, project.Name)
+	if err != nil {
+		return err
+	}
 	for k, volume := range project.Volumes {
 		volume.Labels = volume.Labels.Add(api.VolumeLabel, k)
 		volume.Labels = volume.Labels.Add(api.ProjectLabel, project.Name)
 		volume.Labels = volume.Labels.Add(api.VersionLabel, api.ComposeVersion)
-		err := s.ensureVolume(ctx, volume, project.Name)
+		volumeConfigHash, err := VolumeHash(volume)
+		if err != nil {
+			return err
+		}
+		volume.Labels = volume.Labels.Add(api.ConfigHashLabel, volumeConfigHash)
+
+		err = s.ensureVolume(ctx, volume, project.Name)
+		if err != nil {
+			return err
+		}
+
+		actualVolume, exists := observedVolumesState[k]
+		if !exists {
+			continue
+		}
+		// Check if it's necessary to recreate the volume by comparing
+		// with the current volume config hash
+		shouldRecreateVolume, err := shouldUpdateVolume(volume, actualVolume)
 		if err != nil {
 			return err
 		}
+		if !shouldRecreateVolume {
+			continue
+		}
+		if !recreateVolumes && shouldRecreateVolume {
```

Comment:
we could rely on user interaction to confirm recreation, the way we do on https://github.com/docker/compose/blob/main/pkg/compose/remove.go#L88

Reply:
looks really nice, but I will defer it for a follow-up PR